pkg/datapath/linux/probes/probes.go

```diff
@@ -386,10 +387,145 @@ func HaveBoundedLoops() error {
 }
 
 // HaveFibIfindex checks if kernel has d1c362e1dd68 ("bpf: Always return target
-// ifindex in bpf_fib_lookup") which is 5.10+. This got merged in the same kernel
-// as the new redirect helpers.
-func HaveFibIfindex() error {
-	return features.HaveProgramHelper(ebpf.SchedCLS, asm.FnRedirectPeer)
+// ifindex in bpf_fib_lookup") which is 5.10+ (and backports). It does so by
+// checking that ifindex has been updated as a result of the RIB lookup even
+// though the bpf_fib_lookup failed (no L2 resolution for the NH)
+func HaveFibIfindex() (err error) {
```

Comment:
Good point. I am not really sure. I can tell that is being executed 2x during startup, but I don't know if it's being executed during normal operation.

Let me investigate how could it be cached

Reply:
For caching there is https://go.dev/pkg/sync/?m=old#OnceValues If other functions in the package don't cache you probably also don't have to cache here.

Reply:
So first, other runtime probes are not cached (e.g. [HaveOuterSourceIPSupport()](https://github.com/cilium/cilium/blob/14f0780195f498a2395089c56a871caecfd17f61/pkg/datapath/linux/probes/probes.go#L472) or [HaveOuterSourceIPSupport()](https://github.com/cilium/cilium/blob/3f5e735cdb303cdcba019a02f46cc781006dc386/pkg/datapath/linux/probes/probes.go#L532)), albeit they are not called more than once.

The currrent `HaveFibIfindex()` function is called from three different places:

* Directly from  https://github.com/cilium/cilium/blob/3f5e735cdb303cdcba019a02f46cc781006dc386/daemon/cmd/kube_proxy_replacement.go#L404
* From Unix metrics (`enableIfIndexMetric()`): https://github.com/cilium/cilium/blob/3f5e735cdb303cdcba019a02f46cc781006dc386/pkg/metrics/metrics_unix.go#L29
* From `ExecuteHeaderProbes` [here](https://github.com/cilium/cilium/blob/3f5e735cdb303cdcba019a02f46cc781006dc386/pkg/datapath/linux/probes/probes.go#L720), which then is cached in a boolean flag. `ExecuteHeaderProbes` is later called by the `daemon_main.go`: https://github.com/cilium/cilium/blob/3f5e735cdb303cdcba019a02f46cc781006dc386/daemon/cmd/daemon_main.go#L1275. `.

There is probably a reason (not all daemons require all probes?), can't tell right now, but sounds a bit convoluted.

Side comment, wondering if a singleton pattern for _all_ probes would make sense, so that there is a tight control over _when_ probes run and can be easily and cleanly cached cached.

In any case, a) this doesn't seem anything strictly related to this PR b) to my understanding this doesn't run other than at startup.

Let me know if you think I should address this in this PR, open an issue or s/t.